File: BST/go/bst.go
package main

import (
	"errors"
	"fmt"
	"strconv"
)

type Node struct {
	value int
	left  *Node
	right *Node
}

func (node *Node) Insert(value int) bool {
	if value < node.value {
		if node.left == nil {
			node.left = &Node{value: value}
			return true
		} else {
			return node.left.Insert(value)
		}
	}

	if node.right == nil {
		node.right = &Node{value: value}
		return true
	} else {
		return node.right.Insert(value)
	}
}

func BuildTree(array []int) (Node, error) {
	node := Node{value: array[0]}
	for i := 1; i < len(array); i++ {
		if ok := node.Insert(array[i]); !ok {
			message := fmt.Sprintf("Insertion failed for element %v.", array[i])
			return node, errors.New(message)
		}
	}
	return node, nil
}

func (node Node) Print() string {
	line := ""
	if node.left != nil {
		line += node.left.Print()
	}
	line += strconv.Itoa(node.value) + " "
	if node.right != nil {
		line += node.right.Print()
	}
	return line
}

func (node Node) Sort() (sortedArray []int) {
	if node.left != nil {
		sortedArray = append(sortedArray, node.left.Sort()...)
	}
	sortedArray = append(sortedArray, node.value)
	if node.right != nil {
		sortedArray = append(sortedArray, node.right.Sort()...)
	}
	return sortedArray
}

func main() {
	array := []int{5, 3, 7, 1, 8, 4, 9}
	tree, err := BuildTree(array)
	if err != nil {
		fmt.Print(err)
	}
	result1 := tree.Print()
	fmt.Printf("%v, %T\n", result1, result1)
	result2 := tree.Sort()
	fmt.Printf("%v, %T\n", result2, result2)
}

File: BST/python/bst.py
from __future__ import annotations

from typing import Optional


class Node:
    value: Optional[int]
    left: Optional[Node]
    right: Optional[Node]

    def __init__(self):
        self.value = None
        self.left = None
        self.right = None

    def insert(self, value: int) -> bool:
        """ Insert value into a binary tree. True if succeeded, False otherwise. """
        if self.value is None:
            self.value = value
            self.left = Node()
            self.right = Node()
            return True
        if value < self.value:
            return self.left.insert(value)
        return self.right.insert(value)

    def search_for(self, value: int) -> bool:
        """ Search for value inside a binary tree, True if found False otherwise. """
        if self.value is None:
            return False
        if value == self.value:
            return True
        if value < self.value:
            return self.left.search_for(value)
        if value > self.value:
            return self.right.search_for(value)

    @classmethod
    def build_tree(cls, array: list) -> Node:
        node = Node()
        for value in array:
            if node.insert(value) is False:
                raise ValueError(f"Insertion of element {value} failed.")
        return node

    def sort(self) -> list:
        if not self:
            return []
        sorted_values = []
        if self.left:
            sorted_values.extend(self.left.sort())
        sorted_values.append(self.value)
        if self.right:
            sorted_values.extend(self.right.sort())
        return sorted_values

    def __bool__(self):
        return bool(self.value)

    def __str__(self):
        if not self:
            return ""
        lines = []
        if self.left:
            lines.append(str(self.left))
        lines.append(str(self.value))
        if self.right:
            lines.append(str(self.right))
        return " ".join(lines)


if __name__ == '__main__':
    tree = Node.build_tree([5, 3, 7, 1, 8, 4, 9])
    print(tree.sort())
